### PR TITLE
Remove all deprecations in Kokkos 3

### PR DESCRIPTION
**Pull request by masterleinad**
- Keeps `partiton_master` deprecated.
- Introduces `Kokkos_ENABLE_DISABLED_CODE_4`.

### CODE REVIEW

**Review by dalg24 — COMMENTED** [review]
Do the public header changes in a separate PR or at the very least in a separate commit

File: core/unit_test/default/TestDefaultDeviceTypeInit_1.cpp [hunk 1]
```diff
@@ -1,2 +0,0 @@
-#define KOKKOS_DEFAULTDEVICETYPE_INIT_TEST_01
-#include <TestDefaultDeviceTypeInit.hpp>
```
Comment (ldh4) [inline core/unit_test/default/TestDefaultDeviceTypeInit_1.cpp:2]: Why is this include getting deleted from all `TestDefaultDeviceTypeInit_N.cpp`?

Comment (masterleinad) [reply core/unit_test/default/TestDefaultDeviceTypeInit_1.cpp:2]: These tests are deprecated, see https://github.com/kokkos/kokkos/blob/b2371de0d71192e02d3cdb300ba5cf937d79f4ee/core/unit_test/CMakeLists.txt#L884-L891.

File: cmake/KokkosConfigCommon.cmake.in [hunk 2]
```diff
@@ -14,10 +14,6 @@ ENDFOREACH()
 IF(NOT Kokkos_FIND_QUIETLY)
   MESSAGE(STATUS "Enabled Kokkos devices: ${Kokkos_DEVICES}")
 ENDIF()
-IF(Kokkos_ENABLE_THREADS AND NOT DEPRECATED_CODE_3 IN_LIST Kokkos_OPTIONS)  # for backward compatibility
-  SET(Kokkos_ENABLE_PTHREAD ON)
-  LIST(APPEND Kokkos_DEVICES PTHREAD)
-ENDIF()
```
Comment (dalg24) [inline cmake/KokkosConfigCommon.cmake.in:20]: PTHREAD changes needs to be in a separate commit.
And if we are to remove it then certainly we should drop this
https://github.com/kokkos/kokkos/blob/d19aab9981a2c447e832a7b4eb7b16992328fb14/cmake/kokkos_enable_devices.cmake#L23-L29

File: core/src/Kokkos_Core_fwd.hpp [hunk 3]
```diff
@@ -98,9 +98,6 @@ template <class ExecutionSpace, class MemorySpace>
 struct Device;
 
 // forward declare here so that backend initializer calls can use it.
-#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_3
-struct InitArguments;
-#endif
```
Comment (dalg24) [inline core/src/Kokkos_Core_fwd.hpp:103]: Let's buy people more time on that one

File: core/src/Kokkos_MathematicalFunctions.hpp [hunk 4]
```diff
@@ -105,16 +105,6 @@ using promote_2_t = typename promote_2<T, U>::type;
 #endif
 #endif
 
-#if defined(KOKKOS_ENABLE_DEPRECATED_CODE_3)
```
Comment (dalg24) [inline core/src/Kokkos_MathematicalFunctions.hpp:108]: For the math functions too we should allow for more time to transition

File: core/src/impl/Kokkos_AnalyzePolicy.hpp [hunk 5]
```diff
@@ -210,24 +210,15 @@ constexpr bool warn_if_deprecated(std::false_type) { return true; }
 KOKKOS_DEPRECATED_WITH_COMMENT(
     "Invalid WorkTag template argument in execution policy!!")
 constexpr bool warn_if_deprecated(std::true_type) { return true; }
-#define KOKKOS_IMPL_STATIC_WARNING(...) \
-  static_assert(                        \
-      warn_if_deprecated(std::integral_constant<bool, __VA_ARGS__>()), "")
```
Comment (dalg24) [inline core/src/impl/Kokkos_AnalyzePolicy.hpp:215]: Remove the definition of `warn_if_deprecated` since it won't be used any more

File: core/src/traits/Kokkos_WorkTagTrait.hpp [hunk 6]
```diff
@@ -109,14 +109,9 @@ struct WorkTagTrait : TraitSpecificationBase<WorkTagTrait> {
...
   using trait_matches_specification = std::integral_constant<
-      bool,
-#ifndef KOKKOS_ENABLE_DEPRECATED_CODE_3
-      std::is_empty<T>::value &&
-#else
-      !std::is_void<T>::value &&
-#endif
-          !type_list_any<_trait_matches_spec_predicate<T>::template apply,
-                         _exec_policy_traits_without_work_tag>::value>;
+      bool, std::is_empty<T>::value &&
```
Comment (dalg24) [inline core/src/traits/Kokkos_WorkTagTrait.hpp:112]: Use `std::bool_constant` instead

**Comment by masterleinad** [comment]
Relevant changes in this pull request:
- remove pthread support (except for the trilinos mapping)
- including private headers is an error
- remove deprecations in `conatainers/src/Kokkos_Functional.hpp` (`first_argument` etc.)
- remove deprecated `OffsetView` constructor
- remove deprecated behavior of `UnorderedMap::value_at/key_at`
- remove `CudaSpace::access_error`, `CudaUVMSpace::number_of_allocations`, `CUDA_SAFE_CALL`
- remove `TeamPolicy::vector_length`
- remove `HIPSpace::access_error`, `HIP_SAFE_CALL`
- remove `Kokkos::Experimental:Iterate/Rank/MDRangePolicy` aliases
- remove `Impl::is_array_layout/is_execution_policy/is_execution_space/is_memory_space/is_memory_traits`
- remove `Kokkos::is_space::host_memory_space/host_execution_space/host_mirror_space`, `Kokkos::Impl::is_space`, `Kokkos::Impl::SpaceAccessibility`
- remove `create_mirror_view(space,view,WithoutInitializing)`
- remove `finalize_all`
- `ScopeGuard` warnings are errors
- `partition_master` -> DEPRECATED_CODE_4, remove `partition(...)`, `create_instance`
- remove `ActiveExecutionMemorySpace`, `MemorySpaceAccess`, `verify_space`
- remove `KOKKOS_CONSTEXPR_14`, `KOKKOS_DEPRECATED_TRAILING_ATTRIBUTE`, `KOKKOS_THREAD_LOCAL`
- algorithm/math functions in `Experimental` namespace -> DEPRECATED_CODE_4
- remove `parallel_*` with `string` last.
- remove `Kokkos::is_reducer_type`, array reductions with a raw pointer
- remove ProfilingSection::getName/getSectionID`
- remove `Kokkos::Impl::is_view`
- error out on non-empty `WorkTag`.
- initArguments -> DEPRECATED_CODE_4
- remove `impl/Kokkos_Timer.hpp`

Consider reintroducing DEPRECATED_CODE_3 and use it for the code not removed above but moved to DEPRECATED_CODE_4

File: cmake/compile_tests/pthread.cpp [hunk 7]
```diff
@@ -1,14 +0,0 @@
-#include <pthread.h>
-
-void* kokkos_test(void* args) { return args; }
```
Comment (dalg24) [inline cmake/compile_tests/pthread.cpp:3]: Oversight forgot to remove that file in 9147621eb30f6beb3fd0836db7b86bd77d265d36

**Review by dalg24 — APPROVED** [review]

**Review by crtrott — APPROVED** [review]
